### ISSUE TITLE
swizzled vector implementation is incomplete

### ISSUE THREAD

**Issue by illuhad** [illuhad/hipSYCL#7]
The `detail::vec_swizzle<>` class at the moment lacks most of the operator overloads that `vec<>` provides. This causes most operator expressions to fail if only `vec_swizzle` objects are involved in the operation. The missing operators should therefore be implemented.

**Comment by fodinabor** [comment]
Also.. Why is it, that `.xxx()` is only possible on a `vec<Ty, 3>` and not on a `vec<Ty, 4>`?

**Comment by psalz** [comment]
Yep, we ran into this yesterday as well - the spec isn't super clear on this but I would also assume it should be possible to use any combination of swizzle size and vector size.

**Comment by illuhad** [comment]
Could you please specify where in the spec it is indicated that it should also work with different combinations of swizzle and vector size? For the simple swizzles that we are talking about, it says in the spec in section 4.10.2.1 in the description for `XYZW_SWIZZLE`:

> [...] Where `XYZW_SWIZZLE` is all permutations with repetition of x, y for numElements == 2, x, y,  z for numElements == 3 and x, y, z, w for numElements == 4.   For example `xzyw` and `xyyy`

Permutations implies same size to me, so I believe that the current implementation is in fact correct.

For the `swizzle()` member function the spec does not mention such restrictions and I believe (from the top of my head) that the hipSYCL implementation should be able to handle different combinations of vector size and swizzle index size.

**Comment by fodinabor** [comment]
Hmm... I see your point, but that's not necessarily true, that permutation may only include same size vecs.. other implementations as triSYCL and ComputeCpp at least do support swizzling to smaller vectors:
https://github.com/triSYCL/triSYCL/blob/dc50058eb5d83b8cdddd6a43c994f76ea04a8b0e/include/triSYCL/vec/detail/swiz4.hpp

**Comment by psalz** [comment]
FWIW, ComputeCpp seems to support it!

**Comment by illuhad** [comment]
It could well be that this is a case of a gap between statement and intent in the spec. Let's wait for the result of @fodinabor's request for clarification in the SYCL-Docs repo. If it turns out that it was indeed intended to work with smaller vector types as well, it will be very easy to fix in hipSYCL.
